### PR TITLE
Fix and document Kernel lifecycle handling

### LINKED ISSUE #40: Kernel Scope and Lifetime API
Make the Kernel more robust and tie its lifetime to the kernel id.

- [x] Fix the race condition on restart
- [x] Better document the lifecycle of the Kernel JavaScript object
- [x] Clear Comms and Futures on restart
- [x] Add auto-reconnect logic for failed WebSocket connections

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #40.

### CODE REVIEW

**Comment by blink1073** [comment]
ping @jasongrout, based on our discussions last week.

File: src/kernel.ts [hunk 1]
```diff
@@ -422,6 +459,9 @@ class Kernel implements IKernel {
    * received and validated.
    */
   kernelInfo(): Promise<IKernelInfo> {
+    if (this._status === KernelStatus.Dead) {
+      return Promise.reject(Error('Kernel is not ready to send a message'));
+    }
```
Comment (minrk) [inline src/kernel.ts]: Shouldn't these status checks be in sendKernelMessage (like you already have done in sendShellMessage) to avoid the duplication?

Comment (blink1073) [inline src/kernel.ts]: This one could be, but the others have a more stringent definition of "ready to send".

Comment (minrk) [inline src/kernel.ts]: Ah, I see. Unfortunate, but not a big deal. +1, then.

**Comment by blink1073** [comment]
:beers:

**Pull request merged**